Docs: Update reference for the extensions (#954)
grafana/plugin-tools: docusaurus/docs/how-to-guides/ui-extensions/create-an-extension-point.md

---
id: create-an-extension-point
title: Create an extension point
sidebar_label: Create an extension point
sidebar_position: 10
description: Learn how to provide an extension point so that other applications can contribute their extensions.
keywords:
  - grafana
  - plugins
  - plugin
  - links
  - extensions
  - extension point
  - app plugins
  - apps
---

import ExtensionPoints from '@shared/extension-points.md';

> An _extension point_ is a part of your plugin UI Grafana UI where other plugins can hook in links or React components. You can use them to extend the user experience based on a context exposed by the extension point.

Reviewer: ```suggestion
An _extension point_ is a part of your plugin or UI Grafana UI where other plugins can add links or React components via hooks. You can use them to extend the user experience based on a context exposed by the extension point.
```

> An _extension point_ is a part of your plugin or UI Grafana UI where other plugins can add links or React components via hooks. You can use them to extend the user experience based on a context exposed by the extension point.

Reviewer 2: ```suggestion
An _extension point_ is a part of your plugin UI or Grafana UI where other plugins can add links or React components via hooks. You can use them to extend the user experience based on a context exposed by the extension point.
```

Read more about extensions under [key concepts](../../key-concepts/ui-extensions.md).

<ExtensionPoints/>

## Links

### Best practices for rendering links

- **Make sure your UI handles multiple links** <br /> Multiple plugins may add links to your extension point. Make sure your extension point can handle this and still provide good user experience. See how you can [limit the number of extensions displayed by plugins](#limit-the-number-of-extensions-by-plugins).
- **Share contextual information** <br /> Think about what contextual information could be useful for other plugins and add this to the `context` object. For example, the panel menu extension point shares the `panelId` and the `timeRange`. Note that the `context{}` object always gets frozen before being passed to the links, so it can't be mutated.
- **Avoid unnecessary re-renders** <br />

>   - \*_Static context_

Reviewer 2: ```suggestion
  - **Static context**
```

- **Share contextual information** <br /> Think about what contextual information could be useful for other plugins and pass this as `props` to the components.

### Creating an extension point for components

```tsx
import { usePluginComponents } from '@grafana/runtime';

export const InstanceToolbar = () => {
  // The `extensionPointId` must be prefixed.
  // - Core Grafana -> prefix with "grafana/"
  // - Plugin       -> prefix with "{your-plugin-id}/"
  //
  // This is also what plugins use when they call `addComponent()`
  const extensionPointId = 'myorg-foo-app/toolbar/v1';
  const { components, isLoading } = usePluginComponents({ extensionPointId });

  if (isLoading) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      {/* Loop through the components added by plugins */}
      {components.map(({ id, component: Component }) => (
        <Component key={id} />
      ))}
    </div>
  );
};
```

### Passing data to the components

```tsx
import { usePluginComponents } from '@grafana/runtime';

> // Types for the props (passed as a generic to the following hook)

Reviewer: ```suggestion
// Types for the props (passed as a generic to the hook in the following code block)
```

| **API reference** <br /> - [`addLink()`](../../reference/ui-extensions.md#addlink) - registering a link from a plugin <br /> - [`usePluginLinks()`](../../reference/ui-extensions.md#usepluginlinks) - fetching links registered for an extension point |
| **Component**                                                                                                                                                                                                                                           | Components are React components that can be used to render a custom user experience. <br /><br /> **When to use?** <br /> Use components if you would like to give more freedom for plugins to extend your UI, for example to extend a configuration form with custom parts. <br /><br /> **API reference** <br /> - [`addComponent()`](../../reference/ui-extensions.md#addcomponent) - registering a component from a plugin <br /> - [`usePluginComponents()`](../../reference/ui-extensions.md#useplugincomponents) - fetching components registered for an extension point |

## Links

### Best practices for rendering links

> - **Make sure your UI handles multiple links** <br /> Multiple plugins may add links to your extension point. Make sure your extension point can handle this and still provide good user experience.

Reviewer 3: Done

- **Share contextual information** <br /> Think about what contextual information could be useful for other plugins and add this to the `context` object. For example, the panel menu extension point shares the "panelId" and the "timeRange". Note: the `context{}` object always gets frozen before being passed to the links, so it cannot be mutated.
- **Avoid unnecessary re-renders** <br />

>   - **Static `context`**

Reviewer 2: ```suggestion
  - **Static context*
```
Given the styling oddities of regular and code text

    ```ts
    // Define the `context` object outside of the component if it only has static values
    const context { foo: 'bar' };

    export const InstanceToolbar = () => {
      const { links, isLoading } = usePluginLinks({ extensionPointId, context });
    ```

>   - **Dynamic `context`**

Reviewer 2: ```suggestion
  - **Dynamic context**
```
as above

import ExtensionPoints from '@shared/extension-points.md';

> An _extension point_ is a part of your plugin UI Grafana UI where other plugins can hook in links or React components. You can use them to extend the user experience based on a context exposed by the extension point.

Reviewer 2: we actually need the `or` as it can be from a plugin or from Grafana

Reviewer: yes, good suggestions, done

```tsx
import { usePluginComponents } from '@grafana/runtime';

> // Types for the props (passed as a generic to the following hook)

Reviewer 2: nit: but this isn't referring to the code immediately following the statement, so not sure this makes things clearer

Reviewer: changed it to "hook in the following code block". This is a nit on a nit: doc style guidance is to avoid "above" and "below" in lieu of "previous" and "following"
